File: src/silo/query_engine/query_engine.cpp
#include "silo/query_engine/query_engine.h"

#include <memory>
#include <string>
#include <utility>
#include <vector>

#include <oneapi/tbb/parallel_for.h>
#include <spdlog/spdlog.h>

#include "silo/common/block_timer.h"
#include "silo/common/log.h"
#include "silo/database.h"
#include "silo/query_engine/copy_on_write_bitmap.h"
#include "silo/query_engine/filter/expressions/expression.h"
#include "silo/query_engine/filter/operators/operator.h"
#include "silo/query_engine/query.h"
#include "silo/query_engine/query_result.h"

namespace silo::query_engine {

using filter::expressions::Expression;
using filter::operators::Operator;

QueryResult executeQuery(const Database& database, const std::string& query_string) {
   Query query(query_string);

   SPDLOG_DEBUG("Parsed query: {}", query.filter->toString());

   std::vector<std::string> compiled_queries(database.table->getNumberOfPartitions());
   std::vector<CopyOnWriteBitmap> partition_filters(database.table->getNumberOfPartitions());
   int64_t filter_time;
   {
      const silo::common::BlockTimer timer(filter_time);
      for (size_t partition_index = 0; partition_index != database.table->getNumberOfPartitions();
           partition_index++) {
         std::unique_ptr<Operator> part_filter = query.filter->compile(
            database, database.table->getPartition(partition_index), Expression::AmbiguityMode::NONE
         );
         compiled_queries[partition_index] = part_filter->toString();
         partition_filters[partition_index] = part_filter->evaluate();
      }
   }

   for (uint32_t i = 0; i < database.table->getNumberOfPartitions(); ++i) {
      SPDLOG_DEBUG("Simplified query for partition {}: {}", i, compiled_queries[i]);
   }

   QueryResult query_result;
   int64_t action_time;
   {
      const silo::common::BlockTimer timer(action_time);
      query_result = query.action->executeAndOrder(database, std::move(partition_filters));
   }

   LOG_PERFORMANCE("Query: {}", query_string);
   LOG_PERFORMANCE("Execution (filter): {} microseconds", std::to_string(filter_time));
   LOG_PERFORMANCE("Execution (action): {} microseconds", std::to_string(action_time));

   return query_result;
}

}  // namespace silo::query_engine

File: src/silo/query_engine/query_engine.rs
use std::sync::Arc;
use std::sync::Mutex;
use std::time::Instant;

use crate::common::block_timer::BlockTimer;
use crate::database::Database;
use crate::query_engine::filter_expressions::expression::Expression;
use crate::query_engine::operator_result::OperatorResult;
use crate::query_engine::operators::operator::Operator;
use crate::query_engine::query::Query;
use crate::query_engine::query_result::QueryResult;

pub struct QueryEngine<'a> {
    database: &'a Database,
}

impl<'a> QueryEngine<'a> {
    pub fn new(database: &'a Database) -> Self {
        Self { database }
    }

    pub fn execute_query(&self, query_string: &str) -> QueryResult {
        let query = Query::new(query_string);

        log::debug!("Parsed query: {}", query.filter.to_string());

        let compiled_queries: Arc<Mutex<Vec<String>>> = Arc::new(Mutex::new(vec![String::new(); self.database.partitions.len()]));
        let partition_filters: Arc<Mutex<Vec<OperatorResult>>> = Arc::new(Mutex::new(vec![OperatorResult::default(); self.database.partitions.len()]));

        let filter_time = {
            let timer = Instant::now();
            rayon::scope(|s| {
                for partition_index in 0..self.database.partitions.len() {
                    let compiled_queries = Arc::clone(&compiled_queries);
                    let partition_filters = Arc::clone(&partition_filters);
                    let query_filter = query.filter.clone();
                    let database = self.database;
                    s.spawn(move |_| {
                        let part_filter = query_filter.compile(
                            database,
                            &database.partitions[partition_index],
                            Expression::AmbiguityMode::None,
                        );
                        compiled_queries.lock().unwrap()[partition_index] = part_filter.to_string();
                        partition_filters.lock().unwrap()[partition_index] = part_filter.evaluate();
                    });
                }
            });
            timer.elapsed().as_micros()
        };

        for (i, compiled_query) in compiled_queries.lock().unwrap().iter().enumerate() {
            log::debug!("Simplified query for partition {}: {}", i, compiled_query);
        }

        let action_time = {
            let timer = Instant::now();
            let query_result = query.action.execute_and_order(self.database, partition_filters.lock().unwrap().clone());
            timer.elapsed().as_micros()
        };

        log::info!("Query: {}", query_string);
        log::info!("Execution (filter): {} microseconds", filter_time);
        log::info!("Execution (action): {} microseconds", action_time);

        query_result
    }
}
